javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll

```diff
@@ -202,6 +202,7 @@ private module CachedSteps {
    */
   cached
   predicate returnStep(DataFlow::Node pred, DataFlow::Node succ) {
+    // Note: FlowSteps::CachedSteps::returnStep/2 has copy-paste children
```

Comment:
This comment was quite confusing when doing commit-by-commit review since the function doesn't appear similar to this at all until later in history 😕 

I think we can eliminate the duplication by introducing a shared predicate
```codeql
pragma[inline]
predicate returnStepWithAsyncFlag(DataFlow::Node pred, DataFlow::Node succ, boolean isAsync) {
    // ... bind isAsync to the async-ness of the function being returned from ...
}
```
and then call that from `returnStep/2`:
```codeql
returnStepWithAsyncFlag(pred, succ, false)
```
and from the Promises.qll library
```codeql
returnStepWithAsyncFlag(pred, succ, true)
```

Reply:
In the end the copy-paste was with the exception part of the predicate.  
I've shared the implementation of that part.

Reply:
There is still a slight but of almost copy-pasted code between the predicate and `DataFlow::localFlowStep`, but I think that is OK.